rotkehlchen/tasks/calendar.py

```diff
@@ -109,60 +114,121 @@ def get_history_events(self, event_types: list[tuple[HistoryEventType, HistoryEv
-                    description=description,
-                    counterparty=counterparty,
-                    address=user_address,
-                    blockchain=blockchain,
-                    color=color,
-                    auto_delete=True,
-                ))
 
-            if self.db_calendar.count_reminder_entries(
-                    event_id=(entry_id := calendar_entry.identifier),
-            ) > 0:  # already has a reminder entry
-                return None  # we don't add any new automatic reminders
+        # else calendar entry already exists
+        calendar_entry = calendar_entries['entries'][0]
+        if timestamp != calendar_entry.timestamp:  # update entry if timestamps don't match
+            self.db_calendar.update_calendar_entry(CalendarEntry(
+                identifier=calendar_entry.identifier,
+                name=name,
+                timestamp=timestamp,
+                description=description,
+                counterparty=counterparty,
+                address=address,
+                blockchain=blockchain,
+                color=color,
+                auto_delete=True,
+            ))
+
+        if self.db_calendar.count_reminder_entries(calendar_entry.identifier) > 0:
+            return None  # already has a reminder entry so don't add any new ones
 
-        return entry_id
+        return calendar_entry.identifier
+
+    def create_or_update_calendar_entry_from_event(
+            self,
+            event: 'EvmEvent',
+            name: str,
+            description: str,
+            timestamp: Timestamp,
+            color: 'HexColorCode',
+            counterparty: str,
+    ) -> int | None:
+        """Create calendar entry from an event.
+        Returns the id of the created entry or None if no entry was created.
+        Also may return None in the case the entry was created but aleady has had a reminder set.
+        """
+        assert event.location_label is not None
+        if (
+            (user_address := string_to_evm_address(event.location_label)) not in self.blockchain_accounts.get(  # noqa: E501
+                blockchain := ChainID.deserialize(event.location.to_chain_id()).to_blockchain(),
+            ) or timestamp <= self.current_ts
+        ):
+            return None  # Skip events in the past or from a different address
+
+        return self.create_or_update_calendar_entry(
+            name=name,
+            timestamp=timestamp,
+            description=description,
+            counterparty=counterparty,
+            address=user_address,
+            blockchain=blockchain,
+            color=color,
+        )
+
+    def delete_calendar_entry(
+            self,
+            name: str,
+            counterparty: str,
+            address: ChecksumAddress,
+            blockchain: SupportedBlockchain,
+    ) -> None:
+        """Delete calendar entry and associated reminders if it exists"""
+        if (calendar_entries := self.db_calendar.query_calendar_entry(
```

Comment:
This if and the function call in its if condition statement are the same cop pasta in all the functions. Abstract all these common functionalities properly in a helper function that is used everywhere and stop repeating code